### PR TITLE
Chore/Additional PoX-4 Rust Tests

### PR DESCRIPTION
**Pull request by setzeus**
### Description

This PR introduces three tests for core public PoX-4 functions currently missing unit tests, specifically:

- Delegate-stack-extend
- Stack-increase
- Delegate-stack-increase

All three tests live in .boot/pox_4_tests.rs with a few helpers defined in .boot/contract_tests.rs.

### Applicable issues
N/A, general Argon/PoX-4 workflow.

### Additional info (benefits, drawbacks, caveats)

Increases code coverage for PoX-4. There's a Response(DataResponse...) extract that might end up being refactored for standardization but no immediate drawbacks.

### CODE REVIEW

File: stackslib/src/chainstate/stacks/boot/mod.rs [hunk 1]
```diff
@@ -1824,6 +1824,65 @@ pub mod test {
...
+            POX_4_NAME,
+            "stack-aggregation-commit-indexed",
+            vec![addr_tuple, Value::UInt(amount)],
+        )
+        .unwrap();
+
+        make_tx(key, nonce, 0, payload)
+    }
+
+    pub fn make_pox_4_increase_stx(
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/mod.rs:1870]: better make_pox_4_stack_increase

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 2]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+                PrincipalData::from(key_to_stacks_addr(stacker_key)).into(),
+                Value::UInt(100),
+                pox_addr.clone(),
+                Value::UInt(block_height as u128),
+                Value::UInt(lock_period),
+                signer_key_val.clone(),
+            ],
+        ),
+    ];
+    // (define-public (delegate-stack-stx (stacker principal)
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1765]: This could also go into a make_pox_4_delegate_stack_stx function

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 3]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+    )
+    .expect("No delegation state, delegate-stx failed")
+    .expect_tuple();
+
+    let stacking_state = get_stacking_state_pox_4(
+        &mut peer,
+        &latest_block,
+        &key_to_stacks_addr(stacker_key).to_account_principal(),
+    )
+    .expect("No stacking state, stack-stx failed")
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1786]: stack-stx was never called here.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 4]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+    .expect_tuple();
+
+    let state_signer_key = stacking_state.get("signer-key").unwrap();
+    assert_eq!(state_signer_key.to_string(), signer_key_val.to_string());
+
+    // (define-public (delegate-stack-extend (stacker principal)
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1792]: This comments should be removed

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 5]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+    //     (extend-count uint)
+    //     (pox-addr { version: (buff 1), hashbytes: (buff 32) })
+    //     (signer-key (buff 33)))
+
+    stacker_nonce += 1;
+
+    let mut latest_block = peer.tenure_with_txs(&txs, &mut coinbase_nonce);
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1799]: processing the same txs again?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 6]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+        &latest_block,
+        &key_to_stacks_addr(stacker_key).to_account_principal(),
+    )
+    .expect("No stacking state, stack-stx failed")
+    .expect_tuple();
+
+    let state_signer_key = stacking_state.get("signer-key").unwrap();
+    assert_eq!(state_signer_key.to_string(), signer_key_val.to_string());
+
+    // now stack-extend with a new signer-key
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1811]: now delegate-stack-extend..

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 7]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+    let signer_key_new_val = Value::buff_from(vec![
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1812]: Which values are used here? Are they different from the first one? Difficult to see in the test here.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 8]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+        vec![
+            PrincipalData::from(key_to_stacks_addr(stacker_key)).into(),
+            pox_addr.clone(),
+            signer_key_new_val.clone(),
+            Value::UInt(1),
+        ],
+    )];
+
+    latest_block = peer.tenure_with_txs(&update_txs, &mut coinbase_nonce);
+    let new_stacking_state = get_stacking_state_pox_4(
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1832]: maybe add a comment that here we test stored signer key is actually the signer_key_new_val

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 9]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+    .expect_tuple();
+
+    let state_signer_key = stacking_state.get("signer-key").unwrap();
+    assert_eq!(state_signer_key.to_string(), signer_key_val.to_string());
+
+    stacker_nonce += 1;
+
+    // (define-public (stack-increase (increse-by uint)
+    let stack_increase = make_pox_4_increase_stx(stacker_key, stacker_nonce, min_ustx);
+    let latest_block = peer.tenure_with_txs(&vec![stack_increase], &mut coinbase_nonce);
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1903]: at other places you create a tx array and use it as an argument, here you create the array in the argument. Better do the further.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 10]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+                Value::Principal(stacker_address.clone()),
+                Value::UInt(min_ustx),
+                pox_addr.clone(),
+                Value::UInt(block_height as u128),
+                Value::UInt(lock_period),
+                signer_key_val.clone(),
+            ],
+        ),
+    ];
+    // (define-public (delegate-stack-stx (stacker principal)
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1993]: These comments can be removed

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 11]
```diff
@@ -1701,6 +1707,347 @@ fn delegate_stack_stx_signer_key() {
...
+    );
+
+    let txs = vec![delegate_increase];
+
+    let latest_block = peer.tenure_with_txs(&txs, &mut coinbase_nonce);
+
+    let delegate_transactions =
+        get_last_block_sender_transactions(&observer, delegate_address.into());
+
+    let stacker_locked_amount: u128 = match &delegate_transactions[0].result {
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:2025]: There could be a function to expect a transaction result.

**Review by friedger — COMMENTED** [review]
:+1:

**Review by jferrant — APPROVED** [review]
Looks good :D

File: stackslib/src/chainstate/stacks/boot/pox-4.clar [hunk 12]
```diff
@@ -1144,8 +1144,8 @@
 (define-public (delegate-stack-extend
                     (stacker principal)
                     (pox-addr { version: (buff 1), hashbytes: (buff 32) })
-                    (signer-key (buff 33))
-                    (extend-count uint))
+                    (extend-count uint)
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:1147]: This change should be a different PR related to this issue: https://github.com/stacks-network/stacks-core/issues/4230

**Review by friedger — COMMENTED** [review]
comment related to issue https://github.com/stacks-network/stacks-core/issues/4230

File: stackslib/src/chainstate/stacks/boot/mod.rs [hunk 13]
```diff
@@ -1824,6 +1824,117 @@ pub mod test {
...
+                Value::UInt(extend_count),
+                Value::buff_from(signer_key.to_bytes_compressed()).unwrap(),
+            ],
+        )
+        .unwrap();
+
+        make_tx(key, nonce, 0, payload)
+    }
+
+    pub fn make_pox_4_delegate_increase(
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/mod.rs:1879]: This is about stacking by pool operator, not delegation. Therefore, better `make_pox_4_delegate_stack_increase`

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 14]
```diff
@@ -1629,58 +1630,105 @@ fn delegate_stack_stx_signer_key() {
...
+    )
+    .expect("No delegation state, delegate-stx failed")
+    .expect_tuple();
+
+    let stacking_state = get_stacking_state_pox_4(
+        &mut peer,
+        &latest_block,
+        &key_to_stacks_addr(stacker_key).to_account_principal(),
+    )
+    .expect("No stacking state, stack-stx failed")
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1677]: Why `stack-stx` ? Should be `delegate-stack-stx`

**Review by friedger — COMMENTED** [review]
One test needs to be fixed

File: stackslib/src/chainstate/stacks/boot/pox-4.clar (hunk 12 above)
Comment (setzeus) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:1147]: I already address it here & wrote tests for it accordingly. I'd rather update the description to address #4230 as well then revert the changes across all files. 

That work?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 15]
```diff
@@ -1629,58 +1630,105 @@ fn delegate_stack_stx_signer_key() {
...
+    .expect_tuple();
+
+    let state_signer_key = stacking_state.get("signer-key").unwrap();
+
+    assert_eq!(
+        state_signer_key.to_string(),
+        format!("0x{}", signer_public_key.to_hex())
+    );
```
Comment (MarvinJanssen) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1685]: You can compare the values directly.

**Review by MarvinJanssen — COMMENTED** [review]
Good to have more tests.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar (hunk 12 above)
Comment (friedger) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:1147]: The issue does not contain a preferred solution, still unclear what/whether should be implemented.

File: stackslib/src/chainstate/stacks/boot/mod.rs [hunk 16]
```diff
@@ -1824,6 +1824,117 @@ pub mod test {
...
+    }
+
+    pub fn make_pox_4_delegate_stack_increase(
+        key: &StacksPrivateKey,
+        nonce: u64,
+        stacker: &PrincipalData,
+        pox_addr: PoxAddress,
+        amount: u128,
+    ) -> StacksTransaction {
+        //let addr_tuple = Value::Tuple(pox_addr.as_clarity_tuple().unwrap());
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/mod.rs:1886]: Remove commented code, but add more comments :-)

File: stackslib/src/chainstate/stacks/boot/mod.rs [hunk 17]
```diff
@@ -1824,6 +1824,117 @@ pub mod test {
...
+            POX_4_NAME,
+            "stack-aggregation-commit-indexed",
+            vec![addr_tuple, Value::UInt(amount)],
+        )
+        .unwrap();
+
+        make_tx(key, nonce, 0, payload)
+    }
+
+    pub fn make_pox_4_stack_increase(
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/mod.rs:1922]: nit, the order of the make functions is not clear to me.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 18]
```diff
@@ -1629,58 +1630,105 @@ fn delegate_stack_stx_signer_key() {
...
-            ],
-        ),
-    ];
-    // (define-public (delegate-stack-stx (stacker principal)
-    //                                (amount-ustx uint)
-    //                                (pox-addr { version: (buff 1), hashbytes: (buff 32) })
-    //                                (start-burn-ht uint)
-    //                                (lock-period uint)
-    //                                (signer-key (buff 33)))
+    let delegate_stx: StacksTransaction = make_pox_4_delegate_stx(
```
Comment (friedger) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1708]: I'd love to see human-readable descriptions of the tests in pox 4. It is already difficult enough to understand.

**Review by friedger — COMMENTED** [review]
Apart from #4230, this looks good.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 19]
```diff
@@ -1698,7 +1745,195 @@ fn delegate_stack_stx_signer_key() {
...
+    .expect_tuple();
+
+    let state_signer_key_new = new_stacking_state.get("signer-key").unwrap();
+    assert_eq!(
+        state_signer_key_new.to_string(),
+        format!("0x{}", new_signer_public_key.to_hex())
+    );
+}
+
+#[test]
```
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1782]: It's worth adding a comment here that the reason this test is so spartan is because stack-increase is the same in pox-4 as it is in pox-3.  Same goes for delegate-stack-increase.

**Review by jcnelson — COMMENTED** [review]
Thanks for this @setzeus!  Overall this looks good, but please address the other reviewers' comments. Also, could you please add docstrings to your tests, and comment on what's going on each time you perform a contract-call?  Thanks!

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 20]
```diff
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-use std::collections::{HashMap, HashSet, VecDeque};
+use std::collections::{BTreeMap, HashMap, HashSet, VecDeque};
```
Comment (MarvinJanssen) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:17]: Unused import?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 21]
```diff
@@ -1698,7 +1745,195 @@ fn delegate_stack_stx_signer_key() {
     .expect_tuple();
 
     let state_signer_key = stacking_state.get("signer-key").unwrap();
-    assert_eq!(state_signer_key.to_string(), signer_key_val.to_string());
+    assert_eq!(
+        state_signer_key.to_string(),
+        format!("0x{}", signer_public_key.to_hex())
+    );
```
Comment (MarvinJanssen) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1751]: I think it's better to compare the values directly instead of formatting to string.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 22]
```diff
@@ -1698,7 +1745,195 @@ fn delegate_stack_stx_signer_key() {
...
+        &key_to_stacks_addr(stacker_key).to_account_principal(),
+    )
+    .unwrap()
+    .expect_tuple();
+
+    let state_signer_key_new = new_stacking_state.get("signer-key").unwrap();
+    assert_eq!(
+        state_signer_key_new.to_string(),
+        format!("0x{}", new_signer_public_key.to_hex())
+    );
```
Comment (MarvinJanssen) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1779]: Same as above.

**Review by MarvinJanssen — COMMENTED** [review]
Some small items.

File: stackslib/src/chainstate/stacks/boot/pox-4.clar (hunk 12 above)
Comment (setzeus) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar]: Addressed & reverted change.

**Review by kantai — COMMENTED** [review]
These tests look good to me -- can you just add like a couple sentences of comments for each of the test in `pox_4_tests` that describe the test: the behavior it is trying to test, and then how it tests it?

**Comment by setzeus** [comment]
@kantai refactored with more reader-friendly Alice/Bob context & added comments to the three tests in this PR. 

Once the comments are accepted / up to standard I'll open an issue against the remaining tests in pox_4_tests.rs to update them accordingly.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 23]
```diff
@@ -1694,11 +1673,300 @@ fn delegate_stack_stx_signer_key() {
...
+    let stack_increase =
+        make_pox_4_stack_increase(alice_stacking_private_key, alice_nonce, min_ustx);
+    // Next tx arr includes a stack_increase pox_4 helper found in mod.rs
+    let txs = vec![stack_increase];
+    let latest_block = peer.tenure_with_txs(&txs, &mut coinbase_nonce);
+    let stacker_transactions = get_last_block_sender_transactions(&observer, alice_address);
+
+    let actual_result = stacker_transactions
+        .first()
+        .map(|tx| tx.result.clone())
```
Comment (MarvinJanssen) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1859]: Is the map necessary?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs [hunk 24]
```diff
@@ -1694,11 +1673,300 @@ fn delegate_stack_stx_signer_key() {
...
+    let txs = vec![delegate_increase];
+
+    let latest_block = peer.tenure_with_txs(&txs, &mut coinbase_nonce);
+
+    let delegate_transactions =
+        get_last_block_sender_transactions(&observer, bob_delegate_address.into());
+
+    let actual_result = delegate_transactions
+        .first()
+        .map(|tx| tx.result.clone())
```
Comment (MarvinJanssen) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1952]: Same as above.

**Review by MarvinJanssen — COMMENTED** [review]
Looks good!

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs (hunk 23 above)
Comment (setzeus) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1859]: Yes. Removing it breaks tests.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs (hunk 24 above)
Comment (setzeus) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1952]: Same as above.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs (hunk 23 above)
Comment (MarvinJanssen) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1859]: I mean, `first()` returns an `Option`, since you are just cloning you could perhaps use `cloned()`.

Comment (setzeus) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs]: Ah, yeap, understood & updated to use .cloned() instead.

**Review by friedger — APPROVED** [review]
LGTM
